README.md
SCLAlertView
===========

[![Version](https://img.shields.io/cocoapods/v/SCLAlertView.svg?style=flat)](http://cocoadocs.org/docsets/SCLAlertView/)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)

Animated Alert View written in Swift, which can be used as a `UIAlertView` or `UIAlertController` replacement. Since `UIAlertView` is deprecated and `UIAlertController` only works on iOS 8.x or above, if you have a Swift project where you want to support iOS 7.x too, SCLAlertView is an ideal substitution.

![BackgroundImage](https://raw.githubusercontent.com/vikmeup/SCPopUpView/master/successScreenshot.png)_
![BackgroundImage](https://raw.githubusercontent.com/vikmeup/SCPopUpView/master/editScreenshot.png)

Easy to use
----

### Get Started

```swift
// Get started
SCLAlertView().showInfo("Important info", subTitle: "You are great")
```

### Updating the alert view

```swift
let alertViewResponder: SCLAlertViewResponder = SCLAlertView().showSuccess("Hello World", subTitle: "This is a more descriptive text.")

// Upon displaying, change/close view
alertViewResponder.setTitle("New Title") // Rename title
alertViewResponder.setSubTitle("New description") // Rename subtitle
alertViewResponder.close() // Close view
```

### Alternative alert types

```
SCLAlertView().showError("Hello Error", subTitle: "This is a more descriptive error text.") // Error
SCLAlertView().showNotice("Hello Notice", subTitle: "This is a more descriptive notice text.") // Notice
SCLAlertView().showWarning("Hello Warning", subTitle: "This is a more descriptive warning text.") // Warning
SCLAlertView().showInfo("Hello Info", subTitle: "This is a more descriptive info text.") // Info
SCLAlertView().showEdit("Hello Edit", subTitle: "This is a more descriptive info text.") // Edit
```

### Raw call to showTitle()

```swift
SCLAlertView().showTitle(
    "Congratulations", // Title of view
    subTitle: "Operation successfully completed.", // String of view
    duration: 2.0, // Duration to show before closing automatically, default: 0.0
    completeText: "Done", // Optional button value, default: ""
    style: .success, // Styles - see below.
    colorStyle: 0xA429FF,
    colorTextButton: 0xFFFFFF
)
```

### Controls

#### Custom Appearance

```swift
// SCLAlertView.SCLAppearanc has more than 15 different properties to customize. See below.

let appearance = SCLAlertView.SCLAppearance(
    kTitleFont: UIFont(name: "HelveticaNeue", size: 20)!,
    kTextFont: UIFont(name: "HelveticaNeue", size: 14)!,
    kButtonFont: UIFont(name: "HelveticaNeue-Bold", size: 14)!,
    showCloseButton: false
)

let alert = SCLAlertView(appearance: appearance)
```

#### Add buttons

```swift
let alertView = SCLAlertView()
alertView.addButton("First Button", target:self, selector:Selector("firstButton"))
alertView.addButton("Second Button") {
    print("Second button tapped")
}
alertView.showSuccess("Button View", subTitle: "This alert view has buttons")
```

#### Hide default close button

```swift
let appearance = SCLAlertView.SCLAppearance(
    showCloseButton: false
)
let alertView = SCLAlertView(appearance: appearance)
alertView.showSuccess("No button", subTitle: "You will have hard times trying to close me")
```

#### Hide default close button & a duration to close the alert

```swift
let appearance = SCLAlertView.SCLAppearance(
    showCloseButton: false
)
let alertView = SCLAlertView(appearance: appearance)
alertView.showWarning("No button", subTitle: "Just wait for 3 seconds and I will disappear", duration: 3)
```


#### Hide alert icon

```swift
let appearance = SCLAlertView.SCLAppearance(
    showCircularIcon: false
)
let alertView = SCLAlertView(appearance: appearance)
alertView.showSuccess("No icon", subTitle: "This is a clean alert without Icon!")
```

#### Use a custom icon

```swift
let appearance = SCLAlertView.SCLAppearance(
    showCircularIcon: true
)
let alertView = SCLAlertView(appearance: appearance)
let alertViewIcon = UIImage(named: "IconImage") //Replace the IconImage text with the image name
alertView.showInfo("Custom icon", subTitle: "This is a nice alert with a custom icon you choose", circleIconImage: alertViewIcon)
```


#### Add Text fields

```swift
// Add a text field
let alert = SCLAlertView()
let txt = alert.addTextField(title:"Enter your name")
alert.addButton("Show Name") {
    println("Text value: \(txt.text)")
}
alert.showEdit("Edit View", subTitle: "This alert view shows a text box")
```

#### Use a custom subview instead of a subtitle
```swift
// Example of using the view to add two text fields to the alert
// Create the subview
let appearance = SCLAlertView.SCLAppearance(
    kTitleFont: UIFont(name: "HelveticaNeue", size: 20)!,
    kTextFont: UIFont(name: "HelveticaNeue", size: 14)!,
    kButtonFont: UIFont(name: "HelveticaNeue-Bold", size: 14)!,
    showCloseButton: false
)

// Initialize SCLAlertView using custom Appearance
let alert = SCLAlertView(appearance: appearance)

// Creat the subview
let subview = UIView(frame: CGRectMake(0,0,216,70))
let x = (subview.frame.width - 180) / 2

// Add textfield 1
let textfield1 = UITextField(frame: CGRectMake(x,10,180,25))
textfield1.layer.borderColor = UIColor.greenColor().CGColor
textfield1.layer.borderWidth = 1.5
textfield1.layer.cornerRadius = 5
textfield1.placeholder = "Username"
textfield1.textAlignment = NSTextAlignment.Center
subview.addSubview(textfield1)

// Add textfield 2
let textfield2 = UITextField(frame: CGRectMake(x,textfield1.frame.maxY + 10,180,25))
textfield2.secureTextEntry = true
textfield2.layer.borderColor = UIColor.blueColor().CGColor
textfield2.layer.borderWidth = 1.5
textfield2.layer.cornerRadius = 5
textfield1.layer.borderColor = UIColor.blueColor().CGColor
textfield2.placeholder = "Password"
textfield2.textAlignment = NSTextAlignment.Center
subview.addSubview(textfield2)

// Add the subview to the alert's UI property
alert.customSubview = subview
alert.addButton("Login") {
    print("Logged in")
}

// Add Button with Duration Status and custom Colors
alert.addButton("Duration Button", backgroundColor: UIColor.brownColor(), textColor: UIColor.yellowColor(), showDurationStatus: true) {
    print("Duration Button tapped")
}

alert.showInfo("Login", subTitle: "", duration: 10)
```


#### List of properties to customize

```swift
// Button 
kButtonFont: UIFont                     
buttonCornerRadius : CGFloat            
showCloseButton: Bool                   
kButtonHeight: CGFloat                  

// Circle Image
showCircularIcon: Bool
kCircleTopPosition: CGFloat
kCircleBackgroundTopPosition: CGFloat
kCircleHeight: CGFloat
kCircleIconHeight: CGFloat

// Text
kTitleFont: UIFont
kTitleTop:CGFloat
kTitleHeight:CGFloat
kTextFont: UIFont
kTextHeight: CGFloat
kTextFieldHeight: CGFloat
kTextViewdHeight: CGFloat

// View 
kDefaultShadowOpacity: CGFloat          
kWindowWidth: CGFloat
kWindowHeight: CGFloat
shouldAutoDismiss: Bool // Set this false to 'Disable' Auto hideView when SCLButton is tapped
fieldCornerRadius : CGFloat
contentViewCornerRadius : CGFloat
disableTapGesture: Bool // set this to true if adding tableview to subView
```


### Alert View Styles

```swift
enum SCLAlertViewStyle: Int {
    case success, error, notice, warning, info, edit, wait, question
}
```


### Alert show animation Styles

```swift
// Animation Styles
public enum SCLAnimationStyle {
    case noAnimation, topToBottom, bottomToTop, leftToRight, rightToLeft
}
```


Installation
---

SCLAlertView is available through 

### [CocoaPods](http://cocoapods.org)

To install add the following line to your Podfile:

    pod 'SCLAlertView'

### [Carthage](https://github.com/Carthage/Carthage)

To install add the following line to your Cartfile:

`github "vikmeup/SCLAlertView-Swift" "master"`

Collaboration
---

I tried to build an easy to use API, while beeing flexible enough for multiple variations, but I'm sure there are ways of improving and adding more features, so feel free to collaborate with ideas, issues and/or pull requests.

Incoming improvements
---

- More animations
- Performance tests

Has been developed initially for the [Scroll Feed](https://itunes.apple.com/us/app/scroll-feed/id842422195?ls=1&mt=8) app

- Design [@SherzodMx](https://twitter.com/SherzodMx) Sherzod Max
- Development [@vikmeup](https://twitter.com/vikmeup) Viktor Radchenko
- Improvements by [@bih](http://github.com/bih) Bilawal Hameed, [@rizjoj](http://github.com/rizjoj) Riz Joj


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/vikmeup/SCLAlertView-Swift/tree/HEAD)

[Full Changelog](https://github.com/vikmeup/SCLAlertView-Swift/compare/0.7.0...HEAD)

**Closed issues:**

- Main thread related crash [\#369](https://github.com/vikmeup/SCLAlertView-Swift/issues/369)
- Cannot Change Circular Icon Color [\#365](https://github.com/vikmeup/SCLAlertView-Swift/issues/365)
- Swift 4  [\#359](https://github.com/vikmeup/SCLAlertView-Swift/issues/359)
- Duplicate Entries [\#354](https://github.com/vikmeup/SCLAlertView-Swift/issues/354)
- Close button not visible [\#353](https://github.com/vikmeup/SCLAlertView-Swift/issues/353)
- Please make updates to pod for Swift 4 [\#351](https://github.com/vikmeup/SCLAlertView-Swift/issues/351)
- Xcode 9 beta 4 swift 3.2 [\#346](https://github.com/vikmeup/SCLAlertView-Swift/issues/346)
- Module compile with Swift 3.0.2 cannot be imported in Swift 3.1 [\#331](https://github.com/vikmeup/SCLAlertView-Swift/issues/331)
- What version compatible with Swift 3 [\#330](https://github.com/vikmeup/SCLAlertView-Swift/issues/330)
- EXC\_BAD\_INSTRUCTON error [\#329](https://github.com/vikmeup/SCLAlertView-Swift/issues/329)
- Is there a way to add UIPickerView to the alert? [\#317](https://github.com/vikmeup/SCLAlertView-Swift/issues/317)
- Localize the text? [\#315](https://github.com/vikmeup/SCLAlertView-Swift/issues/315)
- Resizing UITextView [\#306](https://github.com/vikmeup/SCLAlertView-Swift/issues/306)
- Don't show empty space if title is empty. [\#300](https://github.com/vikmeup/SCLAlertView-Swift/issues/300)
- Why is the custom icon cropped to a circle? [\#296](https://github.com/vikmeup/SCLAlertView-Swift/issues/296)
- Error for textfiled empty [\#291](https://github.com/vikmeup/SCLAlertView-Swift/issues/291)
- Add 'showQuestion' as an option [\#279](https://github.com/vikmeup/SCLAlertView-Swift/issues/279)
- Prevent from close when button is tapped [\#277](https://github.com/vikmeup/SCLAlertView-Swift/issues/277)
- Is It Possible To Use A Custom InputView For The Textfields? [\#276](https://github.com/vikmeup/SCLAlertView-Swift/issues/276)
- Set the height of the alert dynamically based on customSubview height [\#274](https://github.com/vikmeup/SCLAlertView-Swift/issues/274)
- value of type 'String' has no member heightWithConstrainedWidth [\#266](https://github.com/vikmeup/SCLAlertView-Swift/issues/266)
- Swift 3 issue even with version 0.7 [\#264](https://github.com/vikmeup/SCLAlertView-Swift/issues/264)
- Bottom Padding too big when setting showCircularIcon to False [\#262](https://github.com/vikmeup/SCLAlertView-Swift/issues/262)
- 300+ issues preventing compilation [\#257](https://github.com/vikmeup/SCLAlertView-Swift/issues/257)
- Add @discardableResult to the declaration of the open function to silence the compiler in swift 3.0 [\#243](https://github.com/vikmeup/SCLAlertView-Swift/issues/243)
- iOS 10 \(Xcode 8 - Swift 3\) [\#238](https://github.com/vikmeup/SCLAlertView-Swift/issues/238)
- Swift 3 [\#236](https://github.com/vikmeup/SCLAlertView-Swift/issues/236)
- Xcode 8 and swift 3 [\#234](https://github.com/vikmeup/SCLAlertView-Swift/issues/234)
- Silly question about default Done button [\#228](https://github.com/vikmeup/SCLAlertView-Swift/issues/228)
- how to change  button border color?? [\#203](https://github.com/vikmeup/SCLAlertView-Swift/issues/203)

**Merged pull requests:**

- Fixed a typo [\#366](https://github.com/vikmeup/SCLAlertView-Swift/pull/366) ([celikomer](https://github.com/celikomer))
- swift 4 update [\#361](https://github.com/vikmeup/SCLAlertView-Swift/pull/361) ([ahmedsafadii](https://github.com/ahmedsafadii))
- Changes to support Swift 3 [\#356](https://github.com/vikmeup/SCLAlertView-Swift/pull/356) ([oswagner](https://github.com/oswagner))
- made changes to be compatible with swift 4.0 and xcode 9 beta 5 [\#350](https://github.com/vikmeup/SCLAlertView-Swift/pull/350) ([fullmetalsheep](https://github.com/fullmetalsheep))
- Prevent from selecting alert message - Update SCLAlertView.swift [\#345](https://github.com/vikmeup/SCLAlertView-Swift/pull/345) ([karek314](https://github.com/karek314))
- Added activity animation as appearance parameter and added button for wait example [\#343](https://github.com/vikmeup/SCLAlertView-Swift/pull/343) ([prsolucoes](https://github.com/prsolucoes))
- Scl alert viewcustom icon [\#338](https://github.com/vikmeup/SCLAlertView-Swift/pull/338) ([rakshithAppaiah](https://github.com/rakshithAppaiah))
- Support horizontal buttons [\#336](https://github.com/vikmeup/SCLAlertView-Swift/pull/336) ([donmichael41](https://github.com/donmichael41))
- Fix top offset when title is empty [\#332](https://github.com/vikmeup/SCLAlertView-Swift/pull/332) ([algrid](https://github.com/algrid))
- Overflow crash fix [\#328](https://github.com/vikmeup/SCLAlertView-Swift/pull/328) ([mvdizel](https://github.com/mvdizel))
- - fix bottom space issue when showCircularIcon = false [\#327](https://github.com/vikmeup/SCLAlertView-Swift/pull/327) ([tomdev-biz](https://github.com/tomdev-biz))
- Fix for bugfix \#320 [\#326](https://github.com/vikmeup/SCLAlertView-Swift/pull/326) ([LordNali](https://github.com/LordNali))
- \[Fix\] Dynamic UITextView Height [\#324](https://github.com/vikmeup/SCLAlertView-Swift/pull/324) ([shaunbevan](https://github.com/shaunbevan))
- \[BugFix\] Int Overflow [\#320](https://github.com/vikmeup/SCLAlertView-Swift/pull/320) ([lastMove](https://github.com/lastMove))
- Update Readme.md : Add Carthage [\#319](https://github.com/vikmeup/SCLAlertView-Swift/pull/319) ([lastMove](https://github.com/lastMove))
- Feature/timeout configuration [\#316](https://github.com/vikmeup/SCLAlertView-Swift/pull/316) ([LordNali](https://github.com/LordNali))
- duration becomes timeout with additional options [\#313](https://github.com/vikmeup/SCLAlertView-Swift/pull/313) ([LordNali](https://github.com/LordNali))

Example/SCLAlertViewExample/ViewController.swift
//
//  ViewController.swift
//  SCLAlertViewExample
//
//  Created by Viktor Radchenko on 6/6/14.
//  Copyright (c) 2014 Viktor Radchenko. All rights reserved.
//

import UIKit
import SCLAlertView


let kSuccessTitle = "Congratulations"
let kErrorTitle = "Connection error"
let kNoticeTitle = "Notice"
let kWarningTitle = "Warning"
let kInfoTitle = "Info"
let kSubtitle = "You've just displayed this awesome Pop Up View"

let kDefaultAnimationDuration = 2.0

class ViewController: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func showSuccess(_ sender: AnyObject) {
		let alert = SCLAlertView()
		_ = alert.addButton("First Button", target:self, selector:#selector(ViewController.firstButton))
		_ = alert.addButton("Second Button") {
			print("Second button tapped")
		}
        _ = alert.showSuccess(kSuccessTitle, subTitle: kSubtitle)
    }
    
    @IBAction func showError(_ sender: AnyObject) {
		_ = SCLAlertView().showError("Hold On...", subTitle:"You have not saved your Submission yet. Please save the Submission before accessing the Responses list. Blah de blah de blah, blah. Blah de blah de blah, blah.Blah de blah de blah, blah.Blah de blah de blah, blah.Blah de blah de blah, blah.Blah de blah de blah, blah.", closeButtonTitle:"OK")
//        SCLAlertView().showError(self, title: kErrorTitle, subTitle: kSubtitle)
    }
    
    @IBAction func showNotice(_ sender: AnyObject) {
        let appearance = SCLAlertView.SCLAppearance(dynamicAnimatorActive: true)
        _ = SCLAlertView(appearance: appearance).showNotice(kNoticeTitle, subTitle: kSubtitle)
    }
    
    @IBAction func showWarning(_ sender: AnyObject) {
        _ = SCLAlertView().showWarning(kWarningTitle, subTitle: kSubtitle)
    }
    
    @IBAction func showInfo(_ sender: AnyObject) {
        _ = SCLAlertView().showInfo(kInfoTitle, subTitle: kSubtitle)
    }
    
	@IBAction func showEdit(_ sender: AnyObject) {
        let appearance = SCLAlertView.SCLAppearance(showCloseButton: true)
        let alert = SCLAlertView(appearance: appearance)
		let txt = alert.addTextField("Enter your name")
        _ = alert.addButton("Show Name") {
			print("Text value: \(txt.text ?? "NA")")
		}
		_ = alert.showEdit(kInfoTitle, subTitle:kSubtitle)
	}
    
    @IBAction func showWait(_ sender: AnyObject) {
        let appearance = SCLAlertView.SCLAppearance(
            showCloseButton: false
        )
        
        let alert = SCLAlertView(appearance: appearance).showWait("Download", subTitle: "Processing...", closeButtonTitle: nil, timeout: nil, colorStyle: nil, colorTextButton: 0xFFFFFF, circleIconImage: nil, animationStyle: SCLAnimationStyle.topToBottom)
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            alert.setSubTitle("Progress: 10%")
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                alert.setSubTitle("Progress: 30%")
                
                DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                    alert.setSubTitle("Progress: 50%")
                    
                    DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                        alert.setSubTitle("Progress: 70%")
                        
                        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                            alert.setSubTitle("Progress: 90%")
                            
                            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                                alert.close()
                            }
                        }
                    }
                }
            }
        }
    }
    
    @IBAction func showCustomSubview(_ sender: AnyObject) {
        // Create custom Appearance Configuration
        let appearance = SCLAlertView.SCLAppearance(
            kTitleFont: UIFont(name: "HelveticaNeue", size: 20)!,
            kTextFont: UIFont(name: "HelveticaNeue", size: 14)!,
            kButtonFont: UIFont(name: "HelveticaNeue-Bold", size: 14)!,
            showCloseButton: false,
            dynamicAnimatorActive: true
        )
        
        // Initialize SCLAlertView using custom Appearance
        let alert = SCLAlertView(appearance: appearance)
        
        // Creat the subview
        let subview = UIView(frame: CGRect(x: 0,y: 0,width: 216,height: 70))
        let x = (subview.frame.width - 180) / 2
        
        // Add textfield 1
        let textfield1 = UITextField(frame: CGRect(x: x,y: 10,width: 180,height: 25))
        textfield1.layer.borderColor = UIColor.green.cgColor
        textfield1.layer.borderWidth = 1.5
        textfield1.layer.cornerRadius = 5
        textfield1.placeholder = "Username"
        textfield1.textAlignment = NSTextAlignment.center
        subview.addSubview(textfield1)
        
        // Add textfield 2
        let textfield2 = UITextField(frame: CGRect(x: x,y: textfield1.frame.maxY + 10,width: 180,height: 25))
        textfield2.isSecureTextEntry = true
        textfield2.layer.borderColor = UIColor.blue.cgColor
        textfield2.layer.borderWidth = 1.5
        textfield2.layer.cornerRadius = 5
        textfield1.layer.borderColor = UIColor.blue.cgColor
        textfield2.placeholder = "Password"
        textfield2.textAlignment = NSTextAlignment.center
        subview.addSubview(textfield2)
        
        // Add the subview to the alert's UI property
        alert.customSubview = subview
        _ = alert.addButton("Login") {
            print("Logged in")
        }
        
        // Add Button with visible timeout and custom Colors
        let showTimeout = SCLButton.ShowTimeoutConfiguration(prefix: "(", suffix: " s)")
        _ = alert.addButton("Timeout Button", backgroundColor: UIColor.brown, textColor: UIColor.yellow, showTimeout: showTimeout) {
            print("Timeout Button tapped")
        }

        let timeoutValue: TimeInterval = 10.0
        let timeoutAction: SCLAlertView.SCLTimeoutConfiguration.ActionType = {
            print("Timeout occurred")
        }
        
        _ = alert.showInfo("Login", subTitle: "", timeout: SCLAlertView.SCLTimeoutConfiguration(timeoutValue: timeoutValue, timeoutAction: timeoutAction))
    }
    
    @IBAction func showCustomAlert(_ sender: AnyObject) {
        let appearance = SCLAlertView.SCLAppearance(
            kTitleFont: UIFont(name: "HelveticaNeue", size: 20)!,
            kTextFont: UIFont(name: "HelveticaNeue", size: 14)!,
            kButtonFont: UIFont(name: "HelveticaNeue-Bold", size: 14)!,
            showCloseButton: false,
            dynamicAnimatorActive: true,
            buttonsLayout: .horizontal
        )
        let alert = SCLAlertView(appearance: appearance)
        _ = alert.addButton("First Button", target:self, selector:#selector(ViewController.firstButton))
        _ = alert.addButton("Second Button") {
            print("Second button tapped")
        }
        
        let icon = UIImage(named:"custom_icon.png")
        let color = UIColor.orange
        
        _ = alert.showCustom("Custom Color", subTitle: "Custom color", color: color, icon: icon!)
    }
	
    @objc func firstButton() {
		print("First button tapped")
	}
}
